query/src/sql/statements/value_source.rs

```diff
@@ -99,29 +89,128 @@ impl ValueSource {
         Ok(DataBlock::create(self.schema.clone(), columns))
     }
 
-    pub async fn parser_read(
-        self,
-        bytes: &[u8],
-        analyzer: ExpressionAnalyzer,
-        ctx: Arc<QueryContext>,
-    ) -> Result<DataBlock> {
-        let values = parse_exprs(bytes, ctx.get_current_session().get_type())?;
-
-        let mut blocks = vec![];
-        for value in values {
-            let block = exprs_to_datablock(value, &analyzer, &self.schema, ctx.clone()).await?;
-            blocks.push(block);
+    /// Parse single row value, like ('111', 222, 1 + 1)
+    async fn parse_next_row<'a>(
+        &self,
+        reader: &mut CpBufferReader<'a>,
+        col_size: usize,
+        desers: &mut [Box<dyn TypeDeserializer>],
+        session_type: &SessionType,
+    ) -> Result<()> {
+        let mut datavalues: Option<Vec<DataValue>> = None;
+
+        let _ = reader.ignore_white_spaces()?;
+        reader.checkpoint();
+
+        // Start of the row --- '('
+        if !reader.ignore_byte(b'(')? {
+            return Err(ErrorCode::BadDataValueType(
+                "Must start with parentheses".to_string(),
+            ));
+        }
+
+        for (col, deser) in desers.iter_mut().enumerate().take(col_size) {
+            if let Some(values) = &datavalues {
+                deser.append_data_value(values[col].clone())?;
+                continue;
+            }
+
+            let _ = reader.ignore_white_spaces()?;
+            let col_end = if col + 1 == col_size { b')' } else { b',' };
+            let result = self.try_deserialize_next_column(reader, deser, col_end);
+
+            // Deserializer and expr-parser both will eat the end ')' of the row.
+            if result.is_err() {
+                skip_to_next_row(reader, 1)?;
+                // Parse from expression and set datavalues
+                let buf = reader.get_checkpoint_buffer();
+                let exprs = parse_exprs(buf, session_type)?;
+                reader.reset_checkpoint();
+
+                let values =
+                    exprs_to_datavalue(exprs, &self.analyzer, &self.schema, self.ctx.clone())
+                        .await?;
+                deser.append_data_value(values[col].clone())?;
+                datavalues = Some(values);
+            }
+        }
+
+        Ok(())
+    }
+
+    fn try_deserialize_next_column(
```

Comment:
Let's just use `deserializer.de_text_quoted(reader).`

If error happens we pop from `[0 - current_idx]`

Reply:
Addressed.